drivers/i2c/i2c_nrfx_twim.c

```diff
@@ -53,17 +59,46 @@ static int i2c_nrfx_twim_transfer(struct device *dev, struct i2c_msg *msgs,
 		}
 
 		nrfx_twim_xfer_desc_t cur_xfer = {
-			.p_primary_buf  = msgs[i].buf,
-			.primary_length = msgs[i].len,
-			.address	= addr,
-			.type		= (msgs[i].flags & I2C_MSG_READ) ?
-					  NRFX_TWIM_XFER_RX : NRFX_TWIM_XFER_TX
+			.address	= addr
 		};
+		if (get_dev_data(dev)->concat_buf_size == 0) {
```

Comment:
I think this whole logic here should be reworked a bit.
A message should be copied to the concatenation buffer in two cases:
- when the previous message was copied (`concat_index != 0`)
- when the current message has no `I2C_MSG_STOP` flag and there is a next message and this next message has no `I2C_MSG_RESTART` flag

When the current message has `I2C_MSG_STOP` flag or there is no next message, the TWIM transfer should occur, from the concatenated buffer (when `concat_index != 0`) or directly from the message buffer. And that's it.

There is no need for this special case if `concat_buf_size == 0`. You check if there is enough space in the concatenation buffer prior to copying a message there. That's enough IMO. Additionally, when there is no concatenation buffer and some messages need to be concatenated, there will be an error reported to the user, what is good (in the current form, the user is not in any way warned that the request may not be performed as requested).